assets/src/edit-story/components/canvas/selectionCanvas.js

```diff
@@ -0,0 +1,149 @@
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useRef } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { useStory } from '../../app';
+import useCanvas from '../canvas/useCanvas';
+import withOverlay from '../overlay/withOverlay';
+import InOverlay from '../overlay';
+
+const LassoMode = {
+	OFF: 0,
+	ON: 1,
+	MAYBE: 2,
+};
+
+const Container = withOverlay( styled.div`
+  width: 100%;
+  height: 100%;
+  user-select: none;
+` );
+
+const Lasso = styled.div`
+  display: none;
+  position: absolute;
+  border: 1px dotted ${ ( { theme } ) => theme.colors.selection };
+  z-index: 1;
+`;
+
+function SelectionCanvas( { children } ) {
+	const {
+		actions: { clearSelection },
+	} = useStory();
+
+	const {
+		state: { pageContainer },
+		actions: { clearEditing, selectIntersection },
+	} = useCanvas();
+
+	const overlayRef = useRef( null );
+	const lassoRef = useRef( null );
+	const offsetRef = useRef( [ 0, 0 ] );
+	const startRef = useRef( [ 0, 0 ] );
+	const endRef = useRef( [ 0, 0 ] );
+	const lassoModeRef = useRef( LassoMode.OFF );
+
+	const getLassoBox = () => {
+		const [ x1, y1 ] = startRef.current;
+		const [ x2, y2 ] = endRef.current;
+		const x = Math.min( x1, x2 );
+		const y = Math.min( y1, y2 );
+		const w = Math.abs( x1 - x2 );
+		const h = Math.abs( y1 - y2 );
+		return [ x, y, w, h ];
+	};
+
+	const updateLasso = () => {
+		const lasso = lassoRef.current;
+		if ( ! lasso ) {
+			return;
+		}
+		const lassoMode = lassoModeRef.current;
+		const [ x, y, w, h ] = getLassoBox();
+		lasso.style.left = `${ x }px`;
+		lasso.style.top = `${ y }px`;
+		lasso.style.width = `${ w }px`;
+		lasso.style.height = `${ h }px`;
+		lasso.style.display = lassoMode === LassoMode.ON ? 'block' : 'none';
+	};
+
+	const onMouseDown = ( evt ) => {
+		clearSelection();
+		clearEditing();
+
+		const overlay = overlayRef.current;
+		let offsetX = 0,
+			offsetY = 0;
+		for ( let offsetNode = overlay; offsetNode; offsetNode = offsetNode.offsetParent ) {
+			offsetX += offsetNode.offsetLeft;
+			offsetY += offsetNode.offsetTop;
+		}
+		const x = evt.pageX - offsetX;
+		const y = evt.pageY - offsetY;
+		offsetRef.current = [ offsetX, offsetY ];
+		startRef.current = [ x, y ];
+		endRef.current = [ x, y ];
+		lassoModeRef.current = LassoMode.MAYBE;
+		updateLasso();
+	};
+
+	const onMouseMove = ( evt ) => {
+		if ( lassoModeRef.current === LassoMode.OFF ) {
+			return;
+		}
+		const [ offsetX, offsetY ] = offsetRef.current;
+		const x2 = evt.pageX - offsetX;
+		const y2 = evt.pageY - offsetY;
+		endRef.current[ 0 ] = x2;
+		endRef.current[ 1 ] = y2;
+		lassoModeRef.current = LassoMode.ON;
+		updateLasso();
+	};
+
+	const onMouseUp = ( ) => {
+		if ( lassoModeRef.current === LassoMode.ON ) {
+			const [ ox, oy, width, height ] = getLassoBox();
+			const x = ox - pageContainer.offsetLeft;
+			const y = oy - pageContainer.offsetTop;
+			clearSelection();
+			clearEditing();
+			selectIntersection( { x, y, width, height } );
```

Comment:
Should the selection perhaps continuously update while drawing the lasso? That's what Keynote does, for example. Makes it easier to see which elements one is going to end up selecting.

Reply:
I'd say "no". Because somewhere, I believe, we wanted <kbd>esc</kbd> to be able to cancel lasso once we support keyboard events. Partially related to #3994. But /cc @samitron7 to clarify.